### ISSUE TITLE
The way that EMA affects the training process

### ISSUE THREAD

**Issue by tserdar** [ultralytics/yolov5#6521]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hey!

I have checked the previous issues but I could not find the specific answer to my question in particular. I need to know how exactly the EMA technique affects the training process. 

Is the model updated by EMA at the end of each epoch i.e. is the loss calculated in the next epoch based on the EMA resulted by the previous epoch during forward-pass? Or EMA is recorded/updated totally separately from the model at the end of each epoch so that it literally has no effect on the model during training and it is only relevant when running validation/test/inference?

Thanks

**Comment by glenn-jocher** [comment]
@rocketsfallonrocketfalls EMA is an Exponential Moving Average of the base model. See https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage

**Comment by tserdar** [comment]
Hey Glenn, 

I am aware what is EMA, thanks. My question was regarding the particular implementation in YOLOv5. 

Could you please tell me if the model updated by EMA after each epoch, or EMA is calculated/updated in a totally separate manner such that it has no effect in model training (e.g. during forward pass for loss calculation etc.)?

**Comment by glenn-jocher** [comment]
@rocketsfallonrocketfalls YOLOv5 EMA is called in train.py and defined here:
https://github.com/ultralytics/yolov5/blob/19e0208fc9e33010717e066f9bd65c27db7c2b5c/utils/torch_utils.py#L295-L329

**Comment by serser** [comment]
Is there an ablation shows the training curve/val mAP of EMA vs. w/o EMA?

**Comment by glenn-jocher** [comment]
@serser We don't have an ablation study showing the training curve/val mAP of EMA vs. w/o EMA specific to YOLOv5. However, EMA is a common technique used in model optimization to improve generalization, and its benefits have been demonstrated in various other frameworks. If you have specific concerns about using EMA in YOLOv5, feel free to share, and I'd be happy to assist further.
